File: horizons_py/horizons/pipelines.py
from __future__ import annotations

from typing import Any, Dict, Optional

from .client import HorizonsClient
from . import models


class PipelinesAPI:
    def __init__(self, client: HorizonsClient) -> None:
        self._client = client

    async def run(self, *, spec: Dict[str, Any], inputs: Optional[Any] = None) -> models.PipelineRun:
        body: Dict[str, Any] = {"spec": spec, "inputs": inputs or {}}
        resp = await self._client._request("POST", "/api/v1/pipelines/run", json=body)
        data = await self._client.json(resp)
        return models.PipelineRun.model_validate(data)

    async def get_run(self, run_id: str) -> models.PipelineRun:
        resp = await self._client._request("GET", f"/api/v1/pipelines/runs/{run_id}")
        data = await self._client.json(resp)
        return models.PipelineRun.model_validate(data)

    async def approve(self, run_id: str, step_id: str) -> Dict[str, Any]:
        resp = await self._client._request(
            "POST", f"/api/v1/pipelines/runs/{run_id}/approve/{step_id}", json={}
        )
        return await self._client.json(resp)

    async def cancel(self, run_id: str) -> Dict[str, Any]:
        resp = await self._client._request("POST", f"/api/v1/pipelines/runs/{run_id}/cancel", json={})
        return await self._client.json(resp)

File: horizons_ts/src/pipelines.ts
import { HorizonsClient } from "./client";
import { PipelineRun } from "./types";

export class PipelinesAPI {
  constructor(private client: HorizonsClient) {}

  async run(input: { spec: Record<string, unknown>; inputs?: unknown }): Promise<PipelineRun> {
    return await this.client.request<PipelineRun>("/api/v1/pipelines/run", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ spec: input.spec, inputs: input.inputs ?? {} }),
    });
  }

  async getRun(runId: string): Promise<PipelineRun> {
    return await this.client.request<PipelineRun>(`/api/v1/pipelines/runs/${runId}`, {
      method: "GET",
    });
  }

  async approve(runId: string, stepId: string): Promise<Record<string, unknown>> {
    return await this.client.request(`/api/v1/pipelines/runs/${runId}/approve/${stepId}`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({}),
    });
  }

  async cancel(runId: string): Promise<Record<string, unknown>> {
    return await this.client.request(`/api/v1/pipelines/runs/${runId}/cancel`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({}),
    });
  }
}

